File: app/src/main/java/com/usher/demo/utils/RxUtil.java
package com.usher.demo.utils;

import android.arch.lifecycle.Lifecycle;
import android.arch.lifecycle.LifecycleOwner;

import com.uber.autodispose.AutoDispose;
import com.uber.autodispose.AutoDisposeConverter;
import com.uber.autodispose.android.lifecycle.AndroidLifecycleScopeProvider;

import io.reactivex.ObservableTransformer;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;

public class RxUtil {
    public static <T> AutoDisposeConverter<T> autoDispose(LifecycleOwner owner) {
        return AutoDispose.autoDisposable(AndroidLifecycleScopeProvider.from(owner, Lifecycle.Event.ON_DESTROY));
    }

    public static <T> ObservableTransformer<T, T> getSchedulerComposer() {
        return upstream -> upstream
                .subscribeOn(Schedulers.io())
                .unsubscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread());
    }
}

File: TwigCodesUI/src/main/java/com/twigcodes/ui/util/RxUtil.kt
package com.twigcodes.ui.util

import androidx.lifecycle.Lifecycle
import androidx.lifecycle.LifecycleOwner
import autodispose2.AutoDispose
import autodispose2.AutoDisposeConverter
import autodispose2.androidx.lifecycle.AndroidLifecycleScopeProvider
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
import io.reactivex.rxjava3.core.ObservableTransformer
import io.reactivex.rxjava3.schedulers.Schedulers
import java.util.concurrent.TimeUnit

object RxUtil {
    @JvmStatic
    fun <T> autoDispose(owner: LifecycleOwner): AutoDisposeConverter<T> = AutoDispose.autoDisposable(AndroidLifecycleScopeProvider.from(owner, Lifecycle.Event.ON_DESTROY))

    @JvmStatic
    fun <T> getSchedulerComposer(): ObservableTransformer<T, T> = ObservableTransformer { upstream ->
        upstream
                .subscribeOn(Schedulers.io())
                .unsubscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
    }

    fun <T> singleClick(): ObservableTransformer<T, T> = ObservableTransformer { upstream -> upstream.throttleFirst(500, TimeUnit.MILLISECONDS) }
}